toolkit/components/filewatcher/NativeFileWatcherLnx.cpp

```diff
@@ -0,0 +1,1073 @@
+#include <cstring>
+#include <errno.h>
+
+#include <iostream>
+
+#include <NativeFileWatcherCommons.h>
+
+namespace mozilla {
+
+namespace moz_filewatcher {
+
+// Global signal counter for events reported by inotify. Used in event handler.
+static volatile sig_atomic_t mEventWaiting;
+
+/**
+ * A structure to hold the information about a single inotify watch descriptor.
+ */
+struct WatchedResourceDescriptor
+{
+    // The path on the file system of the watched resource.
+    nsString mPath;
+
+    // The watched descriptor returned from inotify_add_watch().
+    int mWatchedResourceDescriptor;
+
+    WatchedResourceDescriptor(const nsAString& aPath, const int anHandle)
+        : mPath(aPath)
+        , mWatchedResourceDescriptor(anHandle)
+    {
+    }
+};
+
+/**
+ * This runnable is dispatched from the main thread to get the notifications of
+ * the changes in the watched resources by handling and acting on signals from
+ * the kernel.
+ *
+ * This is accomplished with non-blocking reads from an inotify file descriptor
+ * when the signal handler tied to the inotify file descriptor increments an
+ * event counter. Callbacks are then dispatched depending on the type of event
+ * the inotify_event read from the file descriptor describes.
+ */
+class NativeFileWatcherIOTask : public Runnable
+{
+public:
+    explicit NativeFileWatcherIOTask(int inotifyFileDescriptor)
+        : Runnable("NativeFileWatcherIOTask")
+        , mShuttingDown(false)
+        , mInotifyFileDescriptor(inotifyFileDescriptor)
+    {
+    }
+
+    NS_IMETHOD Run() override;
+    nsresult AddPathRunnableMethod(
+            PathRunnablesParametersWrapper* aWrappedParameters);
+    nsresult RemovePathRunnableMethod(
+            PathRunnablesParametersWrapper* aWrappedParameters);
+    nsresult DeactivateRunnableMethod();
+
+private:
+    // Maintain 2 indexes - one by resource path, one by inotify watch descriptor.
+    // Since our watch descriptor is an int, we use nsUint32HashKey to compute the
+    // hashing key. We need 2 indexes in order to quickly look up the
+    // changed resource in the Worker Thread.
+    nsClassHashtable<nsStringHashKey, WatchedResourceDescriptor>
+    mWatchedResourcesByPath;
+    nsDataHashtable<nsUint32HashKey, WatchedResourceDescriptor*>
+    mWatchedResourcesByHandle;
+
+    // The same callback can be associated to multiple watches so we need to keep
+    // them alive as long as there is a watch using them. We create two hashtables
+    // to map directory names to lists of nsMainThreadPtr<callbacks>.
+    nsClassHashtable<nsStringHashKey, ChangeCallbackArray> mChangeCallbacksTable;
+    nsClassHashtable<nsStringHashKey, ErrorCallbackArray> mErrorCallbacksTable;
+
+    // Other methods need to know that a shutdown is in progress.
+    bool mShuttingDown;
+
+    // Main inotify file descriptor (initialized in NativeFileWatcher's Init())
+    int mInotifyFileDescriptor;
+
+    // Queue for the events read from inotify file descriptor
```

Comment:
As mentioned above, avoid `std::queue`.

Just as importantly, document who's in charge of deallocating the `inotify_event*`.